### PR TITLE
fix(aws-s3) validation for JSON arrays in Upload task and add examples

### PR DESCRIPTION
**Pull request by KhalefAhmed**
## What changes are being made and why?

#297
 fix(aws-s3) validation for JSON arrays in Upload task and add some tests and examples  This PR fixes validation for JSON arrays in the S3 Upload task by adding a check for empty arrays and throwing a clear error message. Previously, empty arrays would fail silently without a helpful error. The PR also adds a new example demonstrating how to upload multiple files to S3 using a YAML Collection directly in the from property.

## How the changes have been QAed?
The changes have been tested localy with multiple flows including the following example that demonstrates using YAML Collection for multiple file uploads and with the tested also with the added examples in `Upload` class:

```yml

id: upload_multiple_files_to_s3_3
namespace: company.team

inputs:
  - id: bucket
    type: STRING
    defaults: my-bucket

tasks:
  - id: download_file1
    type: io.kestra.plugin.core.http.Download
    uri: "https://wri-dataportal-prod.s3.amazonaws.com/manual/global_power_plant_database_v_1_3.zip"

  - id: download_file2
    type: io.kestra.plugin.core.http.Download
    uri: "https://wri-dataportal-prod.s3.amazonaws.com/manual/enhancing-adaptation-ambition-supplementary-materials.zip"

  - id: upload_multiple_to_s3
    type: io.kestra.plugin.aws.s3.Upload
    from:
      - "{{ outputs.download_file1.uri }}"
      - "{{ outputs.download_file2.uri }}"
    key: "data/files/"
    bucket: "{{ inputs.bucket }}"
    region: "{{ secret('AWS_DEFAULT_REGION') }}"
    accessKeyId: "{{ secret('AWS_ACCESS_KEY_ID') }}"
    secretKeyId: "{{ secret('AWS_SECRET_KEY_ID') }}"
```

```yml
id: upload_multiple_files_from_json_array
namespace: company.team

tasks:
  # Download several files
  - id: download_file1
    type: io.kestra.plugin.core.http.Download
    uri: "https://wri-dataportal-prod.s3.amazonaws.com/manual/global_power_plant_database_v_1_3.zip"

  - id: download_file2
    type: io.kestra.plugin.core.http.Download
    uri: "https://wri-dataportal-prod.s3.amazonaws.com/manual/enhancing-adaptation-ambition-supplementary-materials.zip" 

  # Create JSON array and upload files to S3
  - id: upload_files_to_s3
    type: io.kestra.plugin.aws.s3.Upload
    from:  "[\"{{ outputs.download_file1.uri }}\", \"{{ outputs.download_file2.uri }}\"]"
    key: "data/files/"
    bucket: "my-bucket"
    region: "eu-west-3"
    accessKeyId: "test"
    secretKeyId: "test"
    endpointOverride: "http://127.0.0.1:4566"

  # Optional: List uploaded files for verification
  - id: list_uploaded_files
    type: io.kestra.plugin.aws.s3.List
    bucket: "my-bucket"
    prefix: "data/files/"
    region: "eu-west-3"
    accessKeyId: "test"
    secretKeyId: "test"
    endpointOverride: "http://127.0.0.1:4566"
```

### Additional unit tests have been added to verify:
- Proper handling of empty JSON arrays
- Single string inputs
- JSON array inputs

### The output object differs based on the upload type:  

- For single file uploads: Returns Output with direct eTag and versionId properties.
- For multiple file uploads: Returns Output with eTags and versionIds as maps where keys are filenames.

These different return structures need proper documentation in the API reference. Examples of both output types should be included to help users understand the expected response format based on their input type and as I don't know how the documentation is generated at the moment, I've put in a single Output Object with nullables for the maps.

Screenshot 2025-05-06 at 21 38 57

```json
"outputs": {
        "key": "path/to/file",
        "files": {
          "myfile_4697694099743596982.upl": {
            "etag": "\"33ca2f4dfa27ecb503c2c2cc2114362a\"",
            "contentLength": 687
          },
          "myfile2_7595782654353983915.upl": {
            "etag": "\"bb2c187fa6d36d969c136c5145d9e44a\"",
            "contentLength": 1106
          }
        },
        "bucket": "my-bucket"
      }
```

### CODE REVIEW

File: src/main/java/io/kestra/plugin/aws/s3/Upload.java [hunk 1]
```diff
@@ -229,169 +301,242 @@ public Output run(RunContext runContext) throws Exception {
...
-            if (this.metadata != null) {
-                builder.metadata(runContext.render(this.metadata).asMap(String.class, String.class));
+            if (filesToUpload.length == 0) {
+                throw new IllegalArgumentException("No files to upload: the 'from' property contains an empty collection or array");
             }
 
-            if (this.cacheControl != null) {
-                builder.cacheControl(runContext.render(this.cacheControl).as(String.class).orElseThrow());
-            }
+            return filesToUpload.length == 1
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/aws/s3/Upload.java:313]: If both single and multiple upload methods are similar, you can extract this to an `UploadFiles` method.

File: src/main/java/io/kestra/plugin/aws/s3/Upload.java [hunk 2]
```diff
@@ -67,37 +69,107 @@
...
+                    namespace: company.team
+
+                    inputs:
+                      - id: bucket
+                        type: STRING
+                        defaults: my-bucket
+
+                      - id: file_url
+                        type: STRING
+                        defaults: https://wri-dataportal-prod.s3.amazonaws.com/manual/global_power_plant_database_v_1_3.zip
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/aws/s3/Upload.java:82]: You can have some dummy url here.

File: src/main/java/io/kestra/plugin/aws/s3/Upload.java [hunk 3]
```diff
@@ -229,169 +301,242 @@ public Output run(RunContext runContext) throws Exception {
...
+            description = "The name of the bucket where the file(s) were uploaded"
+        )
         private final String bucket;
+
+        @Schema(
+            title = "The S3 object key",
+            description = "The key (path) where the file(s) were uploaded in the bucket"
+        )
         private final String key;
+
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/aws/s3/Upload.java:527]: You can add a inner response class to group uploaded files

**Review by Malaydewangan09 — CHANGES_REQUESTED** [review]
LGTM, minor changes.

File: src/main/java/io/kestra/plugin/aws/s3/Upload.java (hunk 1 above)
Comment (KhalefAhmed) [reply src/main/java/io/kestra/plugin/aws/s3/Upload.java:314]: if we want to have a single method, we'll have 2 additional conditions and these two conditions will be multiplied by the number of files in the case of multiple and we will instantiate variables each time that we're unlikely to use, whether it's the map or the two strings.

```java
private Output uploadFiles(RunContext runContext, S3TransferManager transferManager,
                           String bucket, String baseKey, String[] filesToUpload) throws Exception {
    Map<String, FileInfo> fileInfoMap = new HashMap<>();
    String singleFileETag = null;
    String singleFileVersionId = null;

    for (String renderedFrom : filesToUpload) {
        File tempFile = copyFileToTemp(runContext, renderedFrom);
        String fileName = FilenameUtils.getName(renderedFrom);

        String fileKey = filesToUpload.length == 1 ? baseKey : Path.of(baseKey, fileName).toString();

        PutObjectRequest putObjectRequest = createPutObjectRequest(runContext, bucket, fileKey);

        runContext.logger().debug("Uploading to '{}'", putObjectRequest.key());

        FileUpload upload = transferManager.uploadFile(UploadFileRequest.builder()
            .putObjectRequest(putObjectRequest)
            .source(tempFile)
            .build());

        PutObjectResponse response = upload.completionFuture().get().response();
        
      

        if (filesToUpload.length == 1) {
            singleFileETag = response.eTag();
            singleFileVersionId = response.versionId();
        } else {
            fileInfoMap.put(fileName, FileInfo.builder()
                .eTag(response.eTag())
                .versionId(response.versionId())
                .build());
        }

        recordMetrics(runContext, tempFile);
    }

    if (filesToUpload.length == 1) {
        return Output.builder()
            .bucket(bucket)
            .key(baseKey)
            .eTag(singleFileETag)
            .versionId(singleFileVersionId)
            .build();
    } else {
        return Output.builder()
            .bucket(bucket)
            .key(baseKey)
            .files(fileInfoMap)
            .build();
    }
}
```
Personally, I prefer to have two methods, even if there is more code, but each method has its role.

File: src/main/java/io/kestra/plugin/aws/s3/Upload.java (hunk 2 above)
Comment (KhalefAhmed) [reply src/main/java/io/kestra/plugin/aws/s3/Upload.java:83]: examples with the url coming from datapropal already existed before you want me to modify just mine

Comment (Malaydewangan09) [reply src/main/java/io/kestra/plugin/aws/s3/Upload.java:83]: Ah, Okay
You can keep it as it is.
Thanks!

**Comment by KhalefAhmed** [comment]
thank you for your feedbacks @Malaydewangan09
